Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
++ Priorities will be driven by optimization efforts and requests
++ Priority will be given to intrinsics which have already been implemented for other platforms
++ Priority will be given to intrinsics which are equivalent/similar to those actively used in libraries for other platforms
+
+#### Addition of new intrinsics namespaces
+When an intrinsic namespace is introduced code will likely be implemented on different platforms at different times.
+
+It is explicitly OK for ARM64 to implement a namespace which may be implemented on `ARM32` w/o `ARM32` implementation.
+The opposite is also true.
+
+Any platform which does not implement a namespace, must:
+
++ Return `false` for `IsSupported`
++ `throw` on any unimplemented function
+
+#### Partial implementation of intrinsic namespace
+
++ It is preferred that `IsSupported` represents the state of an entire namespace
++ It is certainly required at time of a release, `IsSupported` represents the state of an entire namespace
++ Allow for writing tests and implementing code, `IsSupported == false` may be used when namespace is partially implemented
+
+#### Addition of new intrinsics to existing namespace
+When an intrinsic namespace is introduced, a best guess will be made about the set of useful intrinsics.
+However, the intrinsics will be added as needed or requested for various optimization efforts.
+
++ The set of intrinsics instructions supported by a namespace should only be allowed
+if all platforms reporting `IsSupported` are supported (at least at time of release)
++ It may be preferable to add new namespaces in cases where this creates significant burden
+
+### Half precision floating point
+
+This document will refer to half precision floating point as `Half`.
+
++ Machine learning and Artificial intelligence often use `Half` type to simplify storage and improve processing time.
++ CoreCLR and `CIL` in general do not have general support for a `Half` type
++ There is an open request to expose `Half` intrinsics
++ ARM64 will introduce `System.Runtime.Intrinsics.Half` to support this request 
++ `Half` will be defined as:
+
+```
+    [StructLayout(LayoutKind.Sequential, Size = 2)]
+    public struct Half : struct {}
+``` 
+
+#### ARMv8 Half precision support
+
+ARMv8 baseline support for `Half` is limited.  The following operations are supported
+
++ Loads and Stores
++ Conversion to/from `Float`
++ Widening from `Vector128<Half>` to two `Vector128<Float>`
++ Narrowing from two `Vector128<Float>` to `Vector128<Float>`
+
+It is presumed that even this minimal support could be helpful
+
+Optional `Half` extensions add more complete support
+
+#### `Half` and ARM64 ABI
+
+The proposed `Half` implementation will treat `Half` as raw bits.  
+It is likely therefore that this will not conform to the ARM64 ABI specifically with respect to the HFA
+calling convention.  This will hinder use of `Half` in `PInvoke` calls.  Resolving this ABI issue will
+be a low priority.
```

Comment:
IMO - This feature should have normal to high priority. If we want to have .NET interoperating with AI/ML it is a must.

Reply:
Low priority is a relative term.  For SIMD, the complxity introduced by ABI compatibility cost 1/2 of the development time.  Implementing internal intrisic functionality is higher priority than ABI compatibility.

That said, I really haven't taken the time to understand Half, HFA and the ABI.  It may just work.

Reply:
```An Homogeneous Floating-point Aggregate (HFA) is an Homogeneous Aggregate with a Fundamental Data Type that is a Floating-Point type and at most four uniquely addressable members```

To make this work imported will need to mark `System.Half` as `TYP_HALF`, so that it can be treated as isFloating().  I am not sure how much work would be required, but it might not be that hard.

I will reword this as a later Milestone, rather than a lower priority.  Some things must work before we can attempt to enable this.